ablySpec/Utilities.swift

```diff
@@ -0,0 +1,137 @@
+                    expect(receivedFoo2).to(equal(123))
+                    expect(receivedBar).to(beNil())
+                    expect(receivedFooOrBar).to(equal(123))
+                    expect(receivedAll).to(equal(123))
+                    
+                    eventEmitter.emit("bar", with:456)
+                    
+                    expect(receivedFoo1).to(equal(123))
+                    expect(receivedFoo2).to(equal(123))
+                    expect(receivedBar).to(equal(456))
+                    expect(receivedFooOrBar).to(equal(456))
+                    expect(receivedAll).to(equal(456))
+                    
+                    eventEmitter.emit("qux", with:789)
+                    
+                    expect(receivedFooOrBar).to(equal(456))
+                    expect(receivedAll).to(equal(789))
+                }
+                
+                it("should only call once listeners once for its event") {
+                    eventEmitter.emit("foo", with: 123)
+
+                    expect(receivedBarOnce).to(beNil())
+                    expect(receivedAllOnce).to(equal(123))
+                    
+                    eventEmitter.emit("bar", with: 456)
+                    
+                    expect(receivedBarOnce).to(equal(456))
+                    expect(receivedAllOnce).to(equal(123))
+
+                    eventEmitter.emit("bar", with: 789)
+                    
+                    expect(receivedBarOnce).to(equal(456))
+                    expect(receivedAllOnce).to(equal(123))
+                }
+                
+                it("should stop receiving events when calling off") {
+                    eventEmitter.offAll(listenerFoo1!)
+                    eventEmitter.emit("foo", with: 123)
+                    
+                    expect(receivedFoo1).to(beNil())
+                    expect(receivedFoo2).to(equal(123))
+                    expect(receivedFooOrBar).to(equal(123))
+                    expect(receivedAll).to(equal(123))
+                    
+                    eventEmitter.off("foo", listener: listenerFooOrBar!)
+                    eventEmitter.emit("foo", with: 456)
+                    
+                    expect(receivedFoo2).to(equal(456))
+                    expect(receivedFooOrBar).to(equal(123))
+                    
+                    eventEmitter.emit("bar", with: 789)
+                    
+                    expect(receivedFooOrBar).to(equal(789))
+                    
+                    eventEmitter.off("foo", listener: listenerAll!)
+                    eventEmitter.emit("foo", with: 111)
+                    
+                    expect(receivedAll).to(equal(789))
+                    
+                    eventEmitter.emit("bar", with: 222)
+                    
+                    expect(receivedAll).to(equal(222))
+                }
+
+                it("should receive only once after calling onAll and then once") {
```

Comment:
Weirdly though, following on from the above example, I did not expect this:

```javascript


myEmitter.removeListener('event', listener);
console.log("> Emitting event third time after remove listener");
myEmitter.emit('event');

/* result from Node.js 
> myEmitter.removeListener('event', listener);
> Emitting event third time after remove listener
> myEmitter.emit('event');
an event occurred!

*/
```

I assumed that calling an `off` handler would match all events, whereas it seems only one is matched.  The Node.js spec says `removeListener will remove, at most, one instance of a listener from the listener array`.

@paddybyers this is worrying as we have not specified the intended behaviour on our 0.8 spec for the EventEmitter. In Ruby I certainly remove all, not sure about the other languages. What do you think we should do, what does Java do?